### PR TITLE
Nonlinear Reservoir HYMOD Integration

### PR DESCRIPTION
**Pull request by jdmattern-noaa**
Integration of nonlinear reservoir into HYMOD.

## Additions

-

## Removals

-

## Changes
#	modified:   models/hymod/include/Hymod.h
#	modified:   models/kernels/Nonlinear_Reservoir.hpp

-

## Testing

1.

## Notes

-

## Todos

-

## Checklist

- [X ] PR has an informative and human-readable title
- [X] Changes are limited to a single goal (no scope creep)
- [X] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [X] Passes all existing automated tests
- [X] Any _change_ in functionality is tested
- [X] New functions are documented (with a description, list of inputs, and expected output)
- [X] Placeholder code is flagged / future todos are captured in comments
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

## Testing checklist (automated report can be put here)

1.

### Target Environment support

- [ ] Linux

### CODE REVIEW

File: models/hymod/include/Hymod.h [hunk 1]
```diff
@@ -127,12 +128,17 @@ class hymod_kernel
         // calculate et
         double et = calc_et(soil_m, et_params);
 
+        double groundwater_excess;  //excess water from groundwater reservoir that can be positive or negative
+        double excess_water;        // excess water from reservoir that can be positive or negative
+
         // get the slow flow output for this time - ks
-        double slow_flow = groundwater.response(slow, dt);
+        double slow_flow = groundwater.response_storage_meters(slow, dt, groundwater_excess);
+        slow_flow += groundwater_excess / dt;
```
Comment (hellkite500) [inline models/hymod/include/Hymod.h]: I'm not sure I like this.  The groundwater reservoir response is slow flow, but the excess isn't nessicarly slow flow, it really should go back to whatever produced slow.  Also, what happens when excess < 0?  At some point, someone has to reconcile the supply/demand mismatch.

File: models/hymod/include/Hymod.h [hunk 2]
```diff
@@ -127,12 +128,17 @@ class hymod_kernel
...
 
         for(unsigned long int i = 0; i < nash_cascade.size(); ++i)
         {
-            runoff = nash_cascade[i].response(runoff, dt);
+            runoff = nash_cascade[i].response_storage_meters(runoff, dt, excess_water);
+            runoff += excess_water / dt;
```
Comment (hellkite500) [inline models/hymod/include/Hymod.h]: Similar to GW comment.

File: models/kernels/Nonlinear_Reservoir.hpp [hunk 3]
```diff
@@ -107,14 +107,25 @@ class Nonlinear_Reservoir
...
 	
+        //Loop through reservoir outlets.
         for (auto& outlet : this->outlets) //pointer to outlets
         {
+            //Calculate outlet velocity.
+            outlet_velocity_meters_per_second = outlet.velocity_meters_per_second(parameters, state);
+
             //Update storage from outlet velocity multiplied by delta time.
-            state.current_storage_height_meters -= outlet.velocity_meters_per_second(parameters, state) * delta_time_seconds;
+            state.current_storage_height_meters -= outlet_velocity_meters_per_second * delta_time_seconds;
```
Comment (hellkite500) [inline models/kernels/Nonlinear_Reservoir.hpp]: Really like the units in these variables, makes dimensional analysis pretty quick!

**Comment by hellkite500** [comment]
Also, in a bit of haste I squashed and merged your old PR, and now I'm seeing the same commits present again.  To avoid having to squash and merge again, you should sync upstream and make a fresh branch for the new changes so we only have the new commits with new changes.

**Comment by jdmattern-noaa** [comment]
> Also, in a bit of haste I squashed and merged your old PR, and now I'm seeing the same commits present again. To avoid having to squash and merge again, you should sync upstream and make a fresh branch for the new changes so we only have the new commits with new changes.

Yeah good point thanks. You showed that to me before, and I will do next time.

**Comment by jdmattern-noaa** [comment]
> Would like to iterate on this just a little bit. Will get @donaldwj to take a look as well. I think the same time consideration we discussed in the LinearReservoir may apply here as well. May need to incorporate the same type of logic for the "internal dt as related to alpha (a)" in the outlet coefficients. In the general case with >1 outlet, I start to wonder if those should/need to be related, or if they can be independent. Thoughts?

Yeah Donald and I spoke about this earlier. He said that passing dt in at response nullifies the need for the internal dt. I will let him clarify.

File: models/hymod/include/Hymod.h (hunk 1 above)
Comment (jdmattern-noaa) [reply models/hymod/include/Hymod.h]: When excess < 0, then it would be subtracted. Do you have any suggestions on what to do with excess?

**Comment by hellkite500** [comment]
What I'm not sure about is it we are adding/subtracting from the correct
place.

File: models/hymod/include/Hymod.h (hunk 1 above)
Comment (ADHydro-devel) [reply models/hymod/include/Hymod.h]: I think that if something happens where the storage goes negative, then it indicates a solution overshoot.   The standard solution is to check if the change in storage this time step is negative and greater than the current storage, then we would make the change in storage equal to the current storage, current storage would be come zero, and we probably want to throw a warning that it happened.

**Comment by hellkite500** [comment]
Replaced by #35.  Will close once that PR is resolved.

File: models/hymod/include/Hymod.h (hunk 1 above)
Comment (jdmattern-noaa) [reply models/hymod/include/Hymod.h:136]: Thanks Fred. I accidentally passed over this comment when you made it. We added boundary checks for min/max and warnings and adjustments to storage to zero if below min and max if exceeding max. We drain the reservoir to min storage in that case by sending the remaining storage to output. We send to excess water back to the flux when over max.

**Comment by hellkite500** [comment]
Done in #35.

**Pull request closed without merging**